### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by fredzqm**
Based on what Zhou Zhou's code base.
I refactored background.js into typescript.
Move firebase to the background.

### CODE REVIEW

File: component/client/backgroundContext.tsx [hunk 1]
```diff
@@ -0,0 +1,4 @@
+const backgroundContext: any = chrome.extension.getBackgroundPage();
```
Comment (fredzqm) [inline component/client/backgroundContext.tsx]: This file groups all reference to background variables.
So it is easy to swap out in the future.

File: component/client/chat/ClientIndependentChatRoom.tsx [hunk 2]
```diff
@@ -24,9 +24,8 @@ class ClientIndependentChatRoom extends React.Component<any, IClientIndependentC
       currentUrl: Constants.NOOP_URL,
       username: Constants.NOOP_USERNAME
     };
-    this.storage = (chrome.extension.getBackgroundPage() as any).BackgroundStorageService;
     this.switchRoom = this.switchRoom.bind(this);
-    this.storage.get("username").then((username) => {
+    storage.get("username").then((username) => {
```
Comment (fredzqm) [inline component/client/chat/ClientIndependentChatRoom.tsx]: storage should be a singleton now.

File: component/client/service/RoomService.tsx [hunk 3]
```diff
@@ -9,7 +11,7 @@ class RoomService {
 
   constructor(url: string, onMessagePosted: (data: IData, from: any) => void) {
     const cleanUrl = url.replace(/[\\.]/g, ",");
-    this.rootRef = firebase.database().ref(cleanUrl);
+    this.rootRef = firebaseDB.ref(cleanUrl);
```
Comment (fredzqm) [inline component/client/service/RoomService.tsx]: refer to the singleton in the background

**Comment by zhoutwo** [comment]
LGTM. Merging.

**Pull request merged**